File: backend/app/models/project.py
from __future__ import annotations

import enum
import uuid
from datetime import UTC, datetime
from decimal import Decimal

from sqlalchemy import DateTime, Enum, ForeignKey, Numeric, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base


class ProjectStatus(str, enum.Enum):
    """Project lifecycle status."""

    ACTIVE = "active"
    ARCHIVED = "archived"


class Project(Base):
    """Project entity — see docs/DATA_MODEL.md for full specification."""

    __tablename__ = "projects"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    color: Mapped[str] = mapped_column(String(7), nullable=False)
    client_name: Mapped[str | None] = mapped_column(String(100), nullable=True)
    hourly_rate: Mapped[Decimal | None] = mapped_column(
        Numeric(10, 2), nullable=True
    )
    status: Mapped[ProjectStatus] = mapped_column(
        Enum(ProjectStatus, name="projectstatus"),
        nullable=False,
        default=ProjectStatus.ACTIVE,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(UTC),
    )

    def __repr__(self) -> str:
        return f"<Project(id={self.id}, name={self.name})>"

File: frontend/src/types/project.ts
export type ProjectStatus = 'active' | 'archived'

export interface Project {
  id: string
  user_id: string
  name: string
  color: string
  client_name: string | null
  hourly_rate: string | null
  status: ProjectStatus
  created_at: string
}

export interface ProjectCreate {
  name: string
  color: string
  client_name?: string | null
  hourly_rate?: string | null
}

export interface ProjectUpdate {
  name?: string
  color?: string
  client_name?: string | null
  hourly_rate?: string | null
  status?: ProjectStatus
}
